frontend/src/domain/entities/vessel/track/index.ts

```diff
@@ -0,0 +1,273 @@
+import { extend } from 'ol/extent'
+import Feature from 'ol/Feature'
+import LineString from 'ol/geom/LineString'
+import Point from 'ol/geom/Point'
+import { transform } from 'ol/proj'
+
+import { getArrowStyle, getCircleStyle, getLineStyle } from '../../../../layers/styles/vesselTrack.style'
+import { calculatePointsDistance, calculateSplitPointCoordinates } from '../../../../utils'
+import { dayjs } from '../../../../utils/dayjs'
+import { Layer } from '../../layers/constants'
+import { OPENLAYERS_PROJECTION, WSG84_PROJECTION } from '../../map'
+import { TRACK_TYPE_RECORD } from './constants'
+
+import type { VesselArrowFeature, VesselLineFeature, VesselPointFeature, VesselId, VesselPosition } from '../types'
+import type { Coordinate } from 'ol/coordinate'
+
+const NUMBER_HOURS_TIME_ELLIPSIS = 4
+const FIRST_POSITION = 0
+
+/**
+ * Get OpenLayers features from vessel positions
+ *
+ * {@see Feature}
+ */
+export function getFeaturesFromPositions(
+  positions: VesselPosition[],
+  vesselId: VesselId
+): (VesselPointFeature | VesselArrowFeature | VesselLineFeature)[] {
+  const hasOnlyOnePosition = positions?.length === 1
+  if (hasOnlyOnePosition) {
+    return getFirstPositionFeature(positions, vesselId)
+  }
+
+  let features: (VesselPointFeature | VesselArrowFeature | VesselLineFeature)[] = []
+  const positionsPointFeatures = buildPointFeatures(positions, vesselId)
+  features = features.concat(positionsPointFeatures)
+
+  const vesselTrackLineFeatures = buildLineStringFeatures(positions, vesselId)
+  features = features.concat(vesselTrackLineFeatures)
+
+  const arrowPointFeatures = buildArrowPointFeatures(vesselTrackLineFeatures, vesselId)
+  features = features.concat(arrowPointFeatures)
+
+  return features
+}
+
+function getFirstPositionFeature(positions: VesselPosition[], vesselId: VesselId) {
+  const position = positions[FIRST_POSITION]
+  if (!position) {
+    throw new Error('No position given')
+  }
+
+  const coordinates = transform([position.longitude, position.latitude], WSG84_PROJECTION, OPENLAYERS_PROJECTION)
+  const feature = buildPointFeature(coordinates, FIRST_POSITION, position, vesselId)
+
+  return [feature]
+}
+
+function buildPointFeatures(positions: VesselPosition[], identity: VesselId): VesselPointFeature[] {
+  return positions
+    .map((position, currentIndex) => {
+      const coordinates = transform([position.longitude, position.latitude], WSG84_PROJECTION, OPENLAYERS_PROJECTION)
+
+      return buildPointFeature(coordinates, currentIndex, position, identity)
+    })
+    .filter((circlePoint): circlePoint is VesselPointFeature => circlePoint !== null)
+}
+
+function buildPointFeature(
+  coordinates: Coordinate,
+  index: number,
+  position: VesselPosition,
+  identity: VesselId
+): VesselPointFeature {
+  const pointFeature = new Feature({
+    geometry: new Point(coordinates)
+  }) as VesselPointFeature
+  pointFeature.name = `${Layer.VESSEL_TRACK.code}:position:${index}`
```

Comment:
Why not a second `options` arg in the constructor?

Reply:
Because it is a Class from OpenLayers ?

Reply:
Théoriquement on peut rajouter des props dans l'objet de Feature, mais elles sont enlevée, je ne comprends pas trop pourquoi